src/projection.jl

```diff
@@ -0,0 +1,144 @@
+    print(io, "ProjectTo{")
+    show(io, T)
+    print(io, "}")
+    if isempty(backing(project))
+        print(io, "()")
+    else
+        show(io, backing(project))
+    end
+end
+
+# fallback (structs)
+function ProjectTo(x::T) where {T}
+    # Generic fallback for structs, recursively make `ProjectTo`s all their fields
+    fields_nt::NamedTuple = backing(x)
+    return ProjectTo{T}(map(ProjectTo, fields_nt))
+end
+function (project::ProjectTo{T})(dx::Tangent) where {T}
+    sub_projects = backing(project)
+    sub_dxs = backing(canonicalize(dx))
+    _call(f, x) = f(x)
+    return construct(T, map(_call, sub_projects, sub_dxs))
+end
+
+# should not work for Tuples and NamedTuples, as not valid tangent types
+function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
+    return throw(
+        ArgumentError("The `x` in `ProjectTo(x)` must be a valid differential, not $x")
+    )
+end
+
+# Generic
+(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
+(::ProjectTo{T})(dx::T) where {T} = dx  # not always true, but we can special case for when it isn't
+(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
+
+# Number
+ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+(::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
+(::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
+
+# Arrays 
+ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
+function (project::ProjectTo{T})(dx::Array) where {T<:Array}
+    _call(f, x) = f(x)
+    return T(map(_call, project.elements, dx))
+end
+function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
+    return T(map(proj -> proj(dx), project.elements))
+end
+(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
+
+# Arrays{<:Number}: optimized case so we don't need a projector per element
+function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
+    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+end
+(project::ProjectTo{<:Array{T}})(dx::Array) where {T<:Number} = project.element.(dx)
+function (project::ProjectTo{<:Array{T}})(dx::AbstractZero) where {T<:Number}
+    return zeros(T, project.size)
+end
+function (project::ProjectTo{<:Array{T}})(dx::Tangent{<:SubArray}) where {T<:Number}
+    return project(dx.parent)
+end
+
+# Diagonal
+ProjectTo(x::T) where {T<:Diagonal} = ProjectTo{T}(; diag=ProjectTo(diag(x)))
+(project::ProjectTo{T})(dx::AbstractMatrix) where {T<:Diagonal} = T(project.diag(diag(dx)))
+(project::ProjectTo{T})(dx::AbstractZero) where {T<:Diagonal} = T(project.diag(dx))
+
+# :data, :uplo fields
+for SymHerm in (:Symmetric, :Hermitian)
+    @eval begin
+        function ProjectTo(x::T) where {T<:$SymHerm}
+            return ProjectTo{T}(; uplo=Symbol(x.uplo), parent=ProjectTo(parent(x)))
+        end
+        function (project::ProjectTo{<:$SymHerm})(dx::AbstractMatrix)
+            return $SymHerm(project.parent(dx), project.uplo)
```

Comment:
This looks a lot like it just applies a Symmetric wrapper, rather than projecting onto the space of symmetric matrices. I think that's wrong, and raised this point on one of the other implementations. Maybe you disagree and can point me to where this was discussed?

Reply:
Earlier comment was https://github.com/JuliaDiff/ChainRulesCore.jl/pull/382#discussion_r658158155

Reply:
Oh, sorry, I meant to comment on this but forgot in the middle of the many small things that came up in the PR. I don't think we are projecting onto the space of symmetric matrices, but rather on the space of `Symmetric` matrices. Which can indeed hold an asymmetric `data` field.

Is there an example that gives an unexpected result? I stared at the finite differencing a little bit and that does seem odd, but not long enough to figure out whether it is a project or finite differencing issue

Reply:
Yes, my Zygote PR has examples. The finite-differencing code was giving bizarre answers and should for now be ignored.

"Many small things" seems accurate, I worry that in rushing to sort them all out, we've lost sight of big-picture questions.